### PR TITLE
[IO-751] api client v2

### PR DESCRIPTION
**Pull request by Nathanjp91**
Adding in the core api client functionality that can make generic http requets. Will be built on for the meta apiclient that is presented to the user and acts as a core base. 
- Generic call method that is invoked by oither calls, get/put/post etc
- Checks for known exceptions
### Changelog
Continuing work on darwin py v2 functionality. Added Core Client

### CODE REVIEW

File: darwin/future/core/client.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    """
+
+    api_key: Optional[str]
+    base_url: str
+    default_team: Optional[Team]
+
+    @validator("base_url")
+    def validate_base_url(cls, v: str) -> str:
+        v = v.strip()
+        if not v.endswith("/"):
```
Comment (owencjones) [inline darwin/future/core/client.py:35]: I see what you mean about Pydantic validators here, they are a bit wordy at times, and the mutation of the value in a validator feels almost kind of dirty.

No change suggested, just WRT prev convo.

File: darwin/future/core/client.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    def post(self, endpoint: str, data: dict) -> dict:
+        return self._generic_call(self.session.post, endpoint, data)
+
+    def delete(self, endpoint: str) -> dict:
+        return self._generic_call(self.session.delete, endpoint)
+
+    def patch(self, endpoint: str, data: dict) -> dict:
+        return self._generic_call(self.session.patch, endpoint, data)
+
+    def _validate_endpoint(self, endpoint: str) -> str:
```
Comment (owencjones) [inline darwin/future/core/client.py:187]: Naming of this one feels off, it's less validating, and more preparing/sanitising/somethingising.  Not a huge deal, consideration point.

File: darwin/future/core/client.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        return endpoint.strip().strip("/")
+
+
+def raise_for_darwin_exception(response: requests.Response) -> None:
+    """Raises an exception if the response is not 200
```
Comment (owencjones) [inline darwin/future/core/client.py:192]: This doesn't raise if the error is in a 500 range, or 318 etc.?  Is that deliberate?

File: pyproject.toml [hunk 4]
```diff
@@ -72,12 +72,19 @@ profile = "black"
 plugins = [
     "pydantic.mypy"
 ]
+
+follow_imports = "silent"
+warn_redundant_casts = true
+warn_unused_ignores = true
+# disallow_any_generics = true
```
Comment (owencjones) [inline pyproject.toml:79]: Just remove the line with this one I think

**Review by owencjones — APPROVED** [review]
Yep, good stuff, few comments to look at.

File: darwin/future/core/client.py (hunk 1 above)
Comment (Nathanjp91) [reply darwin/future/core/client.py:35]: You can seperate them into seperate functions and they become a one line call, I use that elsewhere. Mutation within the validator is the pydantic way, they combine both validation and conversion, if the function throws it's invalid and if it doesn't you mutate into the format you want.

File: darwin/future/core/client.py (hunk 3 above)
Comment (Nathanjp91) [reply darwin/future/core/client.py:192]: It's a good catch on the documentation, it should instead state that it raises for known darwin exceptions rather than any non-200 code. The actual client function that uses it raises_for_status() after checking darwin codes, it'll catch every other code we don't have specific exceptions listed for.

File: darwin/future/core/client.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    api_key: Optional[str]
+    base_url: str
+    default_team: Optional[Team]
+
+    @validator("base_url")
+    def validate_base_url(cls, v: str) -> str:
+        v = v.strip()
+        if not v.endswith("/"):
+            v += "/"
+        assert v.startswith("http") or v.startswith("https"), "base_url must start with http or https"
```
[the remark this answers is not in the document]
Comment (Nathanjp91) [reply darwin/future/core/client.py:37]: Good point on http vs https. Should have caught that myself. 
In terms of urlib there's actually not a validator in that. Closest is a parser that is essentially doing what I'm doing here anyway although a bit more wordily and requiring an import. 

```
from urllib.parse import urlparse
url = urlparse("scheme://netloc/path;parameters?query#fragment")
assert url.scheme
assert url.netloc
assert url.path
```

[the remark this answers is not in the document]
Comment (owencjones) [reply darwin/future/core/client.py:37]: Yeah, I was thinking of `urlparser` - it not necessary obv.

File: darwin/future/core/client.py (hunk 3 above)
Comment (owencjones) [reply darwin/future/core/client.py:192]: 👍🏻
